### PR TITLE
Add solution to ceoi19-amusementpark

### PR DESCRIPTION
**Pull request by rsj001**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/ceoi-19-AmusementPark.mdx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+source: CEOI 2019
+title: 2019 - Amusement Park
+author: Shijie Ren
+---
+
+[CF Editorial](https://codeforces.com/blog/entry/68748)
+
+<Spoiler title="Hint">
+
+We can observe that for each legal proposal , reverse direction of every edges ,
```
Comment (andrewwangva) [inline solutions/ceoi-19-AmusementPark.mdx:12]: Some grammatical errors here. Maybe something like, "We can observe that for each legal proposal, reversing the direction of its edges creates another legal proposal.

File: solutions/ceoi-19-AmusementPark.mdx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+then we'll get another legal proposal.
+
+So the sum of costs will be $m\cdot\texttt{DAGCount}$.
+
+$\texttt{DAGCount}$ represents the number of directed acyclic graphs (DAG)
```
Comment (andrewwangva) [inline solutions/ceoi-19-AmusementPark.mdx:17]: insert "where". "where $\texttt{DAGCount}$ represents the number of directed acyclic graphs"

File: solutions/ceoi-19-AmusementPark.mdx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+---
+
+[CF Editorial](https://codeforces.com/blog/entry/68748)
+
+<Spoiler title="Hint">
+
+We can observe that for each legal proposal , reverse direction of every edges ,
+then we'll get another legal proposal.
+
+So the sum of costs will be $m\cdot\texttt{DAGCount}$.
```
Comment (andrewwangva) [inline solutions/ceoi-19-AmusementPark.mdx:15]: what does this have to do with reversing direction? Isn't it quite obvious that the sum of costs = number of configs * m?

File: solutions/ceoi-19-AmusementPark.mdx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+</Spoiler>
+
+<Spoiler title="Explanation">
+
+Let $\texttt{dp}[S]$ be the number of ways of construct a directed acyclic
+graphs (DAG) with vertex set $S$ , and $\texttt{indepset}[S]$ represents if $S$
+is an independent set.
+
+Since we'll construct a DAG,we can transfer the state by adding new points.
```
Comment (andrewwangva) [inline solutions/ceoi-19-AmusementPark.mdx:28]: fix grammar in this sentence

File: solutions/ceoi-19-AmusementPark.mdx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+Adding 0-out-degree points (or 0-in-degree points is OK) will be easier.
+
+So loop all non-empty subsets of S as points to be added, They could be
+0-out-degree points if and only if they don't have edges with each other in this
+subset. That is, the subset is an independent set.
+
+Also we need to use clusion-exclusion principle.
```
Comment (andrewwangva) [inline solutions/ceoi-19-AmusementPark.mdx:35]: principle of inclusion and exclusion?

File: solutions/ceoi-19-AmusementPark.mdx (hunk 3 above)
Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx:15]: costs of two "opposite" valid proposals = m
so sum of costs = DAGcount / 2 * m
(Forgive my ignorance I don't think it's obvious)

File: solutions/ceoi-19-AmusementPark.mdx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+Loop all non-empty subsets of S as points to be added, points are 0-out-degree
+if and only if they don't have edges with each other in this subset. That is,
+the subset is an independent set.
+
+Also, we need to use the principle of inclusion and exclusion.
+
+$$
+dp[S]=\sum_{P \subset S,P\not=\emptyset,\texttt{indepset}[P]=\texttt{true}}(−1)^{|P|+1}dp[S / P]
+$$
```
Comment (bqi343) [inline solutions/ceoi-19-AmusementPark.mdx:39]: Why is this true? Is it possible to prove this without invoking chromatic polynomials? (https://math.stackexchange.com/questions/2231526/prove-that-number-of-acyclic-orientations-of-a-graph-is-the-chromatic-polynomial)

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/ceoi-19-AmusementPark.mdx (hunk 3 above)
Comment (andrewwangva) [reply solutions/ceoi-19-AmusementPark.mdx]: hmm.. maybe emphasize the two "opposite" valid proposals = m part in your solution

Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx]: Okay

File: solutions/ceoi-19-AmusementPark.mdx (hunk 6 above)
Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx:39]: I am trying to write a (informal) proof, but I encountered a difficulty, maybe it'll be finished in a few days

**Comment by rsj001** [comment]
I don't know if they are correct cuz I am bad at this, plz re-check my explanation

File: solutions/ceoi-19-AmusementPark.mdx (hunk 6 above)
Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx:43]: might be finished

**Comment by rsj001** [comment]
I even confused myself several times lol

**Comment by rsj001** [comment]
my computer was gone this is written on another one today

File: solutions/ceoi-19-AmusementPark.mdx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+Loop all non-empty subsets of S as points to be added, points are 0-out-degree
+if and only if they don't have edges with each other in this subset. That is,
+the subset is an independent set.
+
+<Spoiler title="A Brief Explanation">
+
+Let $A_S$ be the set of DAG configs of the induced subgraph of vertex set $S$.
+
+$$
+A_S = \bigcup_{\varnothing\not=P\subseteq S,P \text{ is an indep set}}A_{S/P} \text{ with edges of points in }S\text{ to }S/P
```
[the remark this answers is not in the document]
Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx:42]: oops I found that I had been using it wrong all the time
including prevoius solutions

File: solutions/ceoi-19-AmusementPark.mdx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+is an independent set.
+
+Since we're constructing a DAG, we can transfer the state by adding new points.
+Adding 0-out-degree points (or 0-in-degree points is OK) will be easier.
+
+Loop all non-empty subsets of S as points to be added, points are 0-out-degree
+if and only if they don't have edges with each other in this subset. That is,
+the subset is an independent set.
+
+<Spoiler title="A Brief Explanation">
```
[the remark this answers is not in the document]
Comment (rsj001) [reply solutions/ceoi-19-AmusementPark.mdx:37]: That's fantastic I found that the size of the sets to be unioned can be added directly at the beginning, but in mine I complicated it cuz I union all configs where P⊆the set of vertices with out-degree 0 at the beginning causes repititions
($A_S$ is an array of numbers? So is it $\sum$ instead of $\cup$?)
and how can I deal with my explanation :eyes:

[the remark this answers is not in the document]
Comment (bqi343) [reply solutions/ceoi-19-AmusementPark.mdx:37]: > ($A_S$ is an array of numbers? So is it $\sum$ instead of $\cup$?)

o yeah

> and how can I deal with my explanation 👀

just use mine if you think it's better?

**Review by bqi343 — APPROVED** [review]
The explanation looks good to me. hint could be reworded but I'll leave that to someone else

**Comment by rsj001** [comment]
.

**Comment by bqi343** [comment]
@andrewwangva @dongliu0426

**Review by dongliu0426 — APPROVED** [review]
